protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,490 @@
+     
+        if (claimRewards) {
+            // can call internal because user reward already updated
+            _getReward(staker, rewardsToAddress);
+        }
+    }
+
+    function _notifyReward(uint256 amount) private {
+        Reward storage rdata = rewardData;
+
+        if (block.timestamp >= rdata.periodFinish) {
+            rdata.rewardRate = uint216(amount / REWARD_DURATION);
+        } else {
+            uint256 remaining = uint256(rdata.periodFinish) - block.timestamp;
+            uint256 leftover = remaining * rdata.rewardRate;
+            rdata.rewardRate = uint216((amount + leftover) / REWARD_DURATION);
+        }
+
+        rdata.lastUpdateTime = uint40(block.timestamp);
+        rdata.periodFinish = uint40(block.timestamp + REWARD_DURATION);
+    }
+
+    function _lastTimeRewardApplicable(uint256 _finishTime) internal view returns (uint256) {
+        if (_finishTime < block.timestamp) {
+            return _finishTime;
+        }
+        return block.timestamp;
+    }
+
+    function _distributeGold() internal {
+        /// @dev no op silent fail if nothing to distribute
+        ITempleGold(address(rewardToken)).mint();
+    }
+
+    function _voteWeight(address _account) private view returns (uint256) {
+        /// @dev Vote weights are always evaluated at the end of last week
+        /// Borrowed from st-yETH staking contract (https://etherscan.io/address/0x583019fF0f430721aDa9cfb4fac8F06cA104d0B4#code)
+        uint256 currentWeek = (block.timestamp / WEEK_LENGTH) - 1;
+        AccountWeightParams storage weight = _weights[_account];
+        uint256 week = uint256(weight.weekNumber);
+        if (week > currentWeek) {
+            weight = _prevWeights[_account];
+        }
+        uint256 t = weight.stakeTime;
+        uint256 updated = weight.updateTime;
+        if (week > 0) {
+            t += (block.timestamp / WEEK_LENGTH * WEEK_LENGTH) - updated;
+        }
+        return _balances[_account] * t / (t + halfTime);
+    }
+
+    function _updateAccountWeight(address _account, uint256 _prevBalance, uint256 _newBalance, bool _increment) private {
+        uint256 currentWeek = block.timestamp / WEEK_LENGTH;
+        uint256 week = 0;
+        uint256 t = 0;
+        uint256 updated = 0;
+        AccountWeightParams storage weight = _weights[_account];
+        (week, t, updated) = _unpackWeight(_account);
+        uint256 _lastShares = _prevBalance;
+        if (week > 0 && currentWeek > week) {
+            _prevWeights[_account] = weight;
+        }
+        if (_newBalance == 0) {
+            t = 0;
+            _lastShares = 0;
+        }
+        if (_lastShares > 0) {
+            t = t + block.timestamp - updated;
+            if (_increment) {
+                // amount has increased, calculate effective time that results in same weight
+                uint256 _halfTime = halfTime;
+                t = _prevBalance * t * _halfTime / (_newBalance * (t + _halfTime) - _prevBalance * t);
```

Comment:
If halfTime is maybe 100, and user tried to 1/100 of his deposited token then `_newBalance * (t + _halfTime) - _prevBalance * t` would become 0